### PR TITLE
Add support for logging service.

### PR DESCRIPTION
**Pull request by llapx**

### CODE REVIEW

**Comment by llapx** [comment]
#https://github.com/ros2/rcl_interfaces/pull/154

**Comment by llapx** [comment]
#https://github.com/ros2/rcl_interfaces/pull/154

**Comment by llapx** [comment]
Rebased to the latest rolling.

**Comment by Barry-Xu-2018** [comment]
This PR depend on https://github.com/ros2/rclcpp/pull/2122

**Comment by Barry-Xu-2018** [comment]
There is another way to avoid changing the existing interface of NodeBase.   
Create a new class NodeBuiltinExecutor. It is initialized at Node construct. `start_parameter_services​`, start_parameter_event_publisher​`, `use_clock_thread` and enable_log_service​​​`  in node options are handled in this class. That is, all builitin pubisher, service, etc are located in class NodeBuiltinExecutor. Node option control if it is created.   
Of course, this leads to big change. So maybe this design is better to be implemented in refactor action instead of implementing this PR.

**Review by fujitatomoya — COMMENTED** [review]
@llapx thanks for working on this, 1st review completed.

File: rclcpp/include/rclcpp/node.hpp [hunk 1]
```diff
@@ -1579,10 +1579,10 @@ class Node : public std::enable_shared_from_this<Node>
 
   rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base_;
   rclcpp::node_interfaces::NodeGraphInterface::SharedPtr node_graph_;
-  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging_;
   rclcpp::node_interfaces::NodeTimersInterface::SharedPtr node_timers_;
   rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics_;
   rclcpp::node_interfaces::NodeServicesInterface::SharedPtr node_services_;
+  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging_;
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/node.hpp:1585]: Yes.

File: rclcpp/src/rclcpp/node.cpp [hunk 2]
```diff
@@ -167,10 +167,14 @@ Node::Node(
...
   node_graph_(new rclcpp::node_interfaces::NodeGraph(node_base_.get())),
-  node_logging_(new rclcpp::node_interfaces::NodeLogging(node_base_.get())),
   node_timers_(new rclcpp::node_interfaces::NodeTimers(node_base_.get())),
   node_topics_(new rclcpp::node_interfaces::NodeTopics(node_base_.get(), node_timers_.get())),
   node_services_(new rclcpp::node_interfaces::NodeServices(node_base_.get())),
+  node_logging_(new rclcpp::node_interfaces::NodeLogging(
+      node_base_,
+      node_services_,
+      options.enable_log_service()
+    )),
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/src/rclcpp/node.cpp:177]: Yes

File: rclcpp/include/rclcpp/node_interfaces/node_base.hpp [hunk 3]
```diff
@@ -98,6 +99,18 @@ class NodeBase : public NodeBaseInterface, public std::enable_shared_from_this<N
     rclcpp::CallbackGroupType group_type,
     bool automatically_add_to_executor_with_node = true) override;
 
+  RCLCPP_PUBLIC
+  rclcpp::CallbackGroup::SharedPtr
+  get_builtin_callback_group() override;
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/node_interfaces/node_base.hpp:104]: Yes

File: rclcpp/src/rclcpp/node.cpp [hunk 4]
```diff
@@ -225,17 +229,19 @@ Node::Node(
     node_topics_->resolve_topic_name("/parameter_events"),
     options.parameter_event_qos(),
     rclcpp::detail::PublisherQosParametersTraits{});
+
+  node_base_->start_builtin_executor_thread();
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/src/rclcpp/node.cpp:233]: Yes. We will change this behavior.

File: rclcpp/src/rclcpp/time_source.cpp [hunk 5]
```diff
@@ -330,7 +324,6 @@ class TimeSource::NodeState final
 
   // Dedicated thread for clock subscription.
   bool use_clock_thread_;
-  std::thread clock_executor_thread_;
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/src/rclcpp/time_source.cpp:333]: Thanks.  
We didn't consider this point.

File: rclcpp/test/rclcpp/test_logging_service.cpp [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  {
+    rclcpp::NodeOptions options = rclcpp::NodeOptions();
+    options.enable_log_service(true);
+    node = std::make_shared<rclcpp::Node>("test_logging_service", "/ns", options);
+  }
+
+  rclcpp::Node::SharedPtr node;
+};
+
+TEST_F(TestLoggingService, get_logger_levels) {
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_logging_service.cpp:51]: Yeah.

File: rclcpp/test/rclcpp/test_logging_service.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    auto result_get = result.get();
+    ASSERT_EQ(result_get->results.size(), 1u);
+    ASSERT_TRUE(result_get->results[0].successful);
+  }
+
+  {
+    // set log level with wrong value
+    auto request = std::make_shared<rcl_interfaces::srv::SetLoggerLevels::Request>();
+    auto level = rcl_interfaces::msg::LoggerLevel();
+    level.name = "/ns/test_logging_service";
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_logging_service.cpp:92]: Yeah

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 since we have not received the response to https://github.com/ros2/rclcpp/pull/2122#issuecomment-1479957010, probably we can keep the current implementation with this PR, which is `sharing single threaded executor within time source and logging level service if enabled`. what do you think?

@iuhilnehc-ynos please chime in if you have comments on this.

thanks,

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya 

> since we have not received the response to https://github.com/ros2/rclcpp/pull/2122#issuecomment-1479957010, probably we can keep the current implementation with this PR, which is sharing single threaded executor within time source and logging level service if enabled. what do you think?

Yes. But I refactor code to avoid many changes on existing interface.  
https://github.com/Barry-Xu-2018/rclcpp/pull/3
For the first version ([f4199fe6613734f4b64934ac654e6176a1c75ee8](https://github.com/Barry-Xu-2018/rclcpp/pull/3/commits/f4199fe6613734f4b64934ac654e6176a1c75ee8)), I used NodeBuiltinExecutor to launch logger service and clock thread. But find time_source has problem for some case.  
After discussing with @iuhilnehc-ynos, we think we should use NodeBuiltinExecutor only for logger service (Otherwise, there are many changes for time_source).
We can create new PR to refactor time_source to use NodeBuiltinExecutor.   
What do you think ?

**Comment by Barry-Xu-2018** [comment]
https://github.com/ros2/rclcpp/pull/2122/commits/0074d6d752a905d9d6b6718ef0b69de63964e121 introduce a new class `NodeBuiltinExecutor`.  
NodeBuiltinExecutor take charge of running all internal(built-In) services, subscription, etc.  
Now only logger service is launched.   
I don't want to have too many changes in this PR which isn't related to logger service.  
Next, clock thread and related work in time source will be moved to this class.

**Comment by Barry-Xu-2018** [comment]
> in addition to https://github.com/ros2/rclcpp/pull/2122#issuecomment-1479957010, we can just rely on user thread to spin just like parameter services ParameterService do. This can be good enough to handle the logging services in the node. I did not think we had discuss on this before, but sounds good enough to me. what do you think?

Yes. We can put logger service in user thread like parameter services.  
From our previous consideration, we want to move all internal services (such as parameter service, clock service, log service, etc) from user thread to internal thread. So we created NodeBuiltinExecutor.  
Of course, it's not appropriate to include such changes for this purpose in this PR.

**Comment by fujitatomoya** [comment]
> Of course, it's not appropriate to include such changes for this purpose in this PR.

I think this can be handles as different issue and PRs, but currently i see that user thread could be most appropriate.

**Comment by Barry-Xu-2018** [comment]
I updated codes.  
@fujitatomoya @iuhilnehc-ynos please review again.

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18461/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13007/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19154/)

**Review by iuhilnehc-ynos — APPROVED** [review]

**Comment by Barry-Xu-2018** [comment]
Rebase was done.

**Review by wjwwood — APPROVED** [review]
Lgtm, I was iffy about the service interface being passed separately from the constructor but on reflection I think that's actually for the best since the feature is optional.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18582/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13124/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19319/)

**Comment by clalancette** [comment]
This needed a rebase, which I've done now.  I think it also needs an updated .repos file, which I've also made.  Here's another CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18583/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13125/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19320/)

**Comment by clalancette** [comment]
(also, Rpr is going to fail since we haven't released `rcl_interfaces` yet; we'll do that tomorrow)

**Comment by clalancette** [comment]
@fujitatomoya @llapx @Barry-Xu-2018 So unfortunately it looks like something is going on with Windows here.  In particular, some of the tests are crashing that didn't crash in the nightlies, and didn't seem to crash with the other code that went in today.  I honestly don't know what is going on here, but I don't think we should put this in like it is.  Can you please take a look and try to see what is happening?

**Comment by Barry-Xu-2018** [comment]
@clalancette Okay. I will check failed test cases.

**Comment by fujitatomoya** [comment]
actually there are 5 test failure, all of them looks not related to our PRs, but hard to guarantee that at this very last moment.
@Barry-Xu-2018 any thoughts?

starting windows CI again:
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19327/)

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya 
Yeah. These failed tests are all the program died or did not respond.  
The logger service isn't enabled, change of this PR should not affect test result.    
So I suspect it's an issue with the CI system. We have to run it again.

**Comment by fujitatomoya** [comment]
@clalancette @Yadunund 

in our opinion, that is so unlikely those are related to our PR, since our code is not even enabled. (builtin logging level service is opt-in.)
so I would say that we can merge this, and we will be there if anything happens to our code as we do.

on the other hand, we understand that you do not take this at very last moment unless CI is all green. (CI windows has been restarted, but will not be completed in time.)

**Pull request merged**